pkg/runner/test_runner.go

```diff
@@ -6,282 +6,778 @@
+	for i, sb := range t.TestStepsBundles {
+		tr.steps = append(tr.steps, &stepState{
+			stepIndex: i,
+			sb:        sb,
+			inCh:      make(chan *target.Target),
+			outCh:     make(chan *target.Target),
+			errCh:     make(chan cerrors.TargetError),
+			ev: storage.NewTestEventEmitter(testevent.Header{
+				JobID:         jobID,
+				RunID:         runID,
+				TestName:      t.Name,
+				TestStepLabel: sb.TestStepLabel,
+			}),
+			tgtDone: make(map[*target.Target]bool),
+			log:     logging.AddField(tr.log, "step", sb.TestStepLabel),
+		})
+	}
+
+	// Set up the targets
+	tr.targets = make(map[string]*targetState)
+	// If we have target state to resume, do it now.
+	if len(resumeState) > 0 {
+		tr.log.Debugf("Attempting to resume from state: %s", string(resumeState))
+		var rs resumeStateStruct
+		if err := json.Unmarshal(resumeState, &rs); err != nil {
+			return nil, fmt.Errorf("invalid resume state: %w", err)
+		}
+		if rs.Version != resumeStateStructVersion {
+			return nil, fmt.Errorf("incompatible resume state version %d (want %d)",
+				rs.Version, resumeStateStructVersion)
+		}
+		if rs.JobID != jobID {
+			return nil, fmt.Errorf("wrong resume state, job id %d (want %d)", rs.JobID, jobID)
+		}
+		tr.targets = rs.Targets
+	}
+	// Initialize remaining fields of the target structures,
+	// build the map and kick off target processing.
+	for _, tgt := range targets {
+		ts := tr.targets[tgt.ID]
+		if ts == nil {
+			ts = &targetState{}
+		}
+		ts.tgt = tgt
+		ts.resCh = make(chan error)
+		tr.mu.Lock()
+		tr.targets[tgt.ID] = ts
+		tr.mu.Unlock()
+		tr.targetsWg.Add(1)
+		go tr.targetRunner(targetCtx, stepCtx, ts)
+	}
+
+	// Run until no more progress can be made.
+	runErr := tr.runMonitor()
+	if runErr != nil {
+		tr.log.Errorf("monitor returned error: %q, canceling", runErr)
+		stepCancel()
+	}
+
+	// Wait for step runners and readers to exit.
+	if err := tr.waitStepRunners(ctx); err != nil {
+		tr.log.Errorf("step runner error: %q, canceling", err)
+		stepCancel()
+		if runErr == nil {
+			runErr = err
+		}
+	}
+	// There will be no more results, reel in all the target runners (if any).
+	tr.log.Debugf("waiting for target runners to finish")
+	targetCancel()
+	tr.targetsWg.Wait()
+
+	// Examine the resulting state.
+	tr.log.Debugf("leaving, err %v, target states:", runErr)
+	tr.mu.Lock()
+	defer tr.mu.Unlock()
+	resumeOk := (runErr == nil)
+	var inFlightTargets []*targetState
+	for i, tgt := range targets {
+		ts := tr.targets[tgt.ID]
+		stepErr := tr.steps[ts.CurStep].runErr
+		tr.log.Debugf("  %d %s %v", i, ts, stepErr)
+		if ts.CurPhase == targetStepPhaseRun {
+			inFlightTargets = append(inFlightTargets, ts)
+			if stepErr != statectx.ErrPaused {
+				resumeOk = false
+			}
+		}
+	}
+	tr.log.Debugf("- %d in flight, ok to resume? %t", len(inFlightTargets), resumeOk)
+	tr.log.Debugf("step states:")
+	for i, ss := range tr.steps {
+		tr.log.Debugf("  %d %s %t %t %v", i, ss, ss.stepRunning, ss.readerRunning, ss.runErr)
+	}
```

Comment:
Better to use structured logs. See `WithField`/`WithFields`. It allows to perform smarter log-based analytics afterwards.